daffodil-runtime1/src/main/scala/org/apache/daffodil/dpath/DFDLXFunctions.scala

```diff
@@ -97,7 +97,59 @@ case class DFDLXRightShift(recipes: List[CompiledDPath], argType: NodeInfo.Kind)
     }
   }
 }
+case class DFDLXBitAnd(recipes: List[CompiledDPath], argType: NodeInfo.Kind) extends FNTwoArgs(recipes) {
+  override def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+    argType match {
+      case NodeInfo.Long => arg1.getLong & arg2.getLong
+      case NodeInfo.Int => arg1.getInt & arg2.getInt
+      case NodeInfo.Short => (arg1.getShort & arg2.getShort).toShort
+      case NodeInfo.Byte => (arg1.getByte & arg2.getByte).toByte
+      case NodeInfo.UnsignedLong => arg1.getBigInt.and(arg2.getBigInt)
+      case NodeInfo.UnsignedInt => arg1.getLong & arg2.getLong
+      case NodeInfo.UnsignedShort => arg1.getInt & arg2.getInt
+      case NodeInfo.UnsignedByte => (arg1.getShort & arg2.getShort).toShort
+      //$COVERAGE-OFF$
+      case _ => Assert.invariantFailed(s"dfdlx:bitAND not supported")
+      // $COVERAGE-ON$
+    }
+  }
+}
 
+case class DFDLXBitOr(recipes: List[CompiledDPath], argType: NodeInfo.Kind) extends FNTwoArgs(recipes) {
+  override def computeValue(arg1: DataValuePrimitive, arg2: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+    argType match {
+      case NodeInfo.Long => arg1.getLong | arg2.getLong
+      case NodeInfo.Int => arg1.getInt | arg2.getInt
+      case NodeInfo.Short => (arg1.getShort | arg2.getShort).toShort
+      case NodeInfo.Byte => (arg1.getByte | arg2.getByte).toByte
+      case NodeInfo.UnsignedLong => arg1.getBigInt.or(arg2.getBigInt)
+      case NodeInfo.UnsignedInt => arg1.getLong | arg2.getLong
+      case NodeInfo.UnsignedShort => arg1.getInt | arg2.getInt
+      case NodeInfo.UnsignedByte => (arg1.getShort | arg2.getShort).toShort
+      //$COVERAGE-OFF$
+      case _ => Assert.invariantFailed(s"dfdlx:bitOR not supported")
+      // $COVERAGE-ON$
+    }
+  }
+}
+
+case class DFDLXBitNot(recipes: CompiledDPath, argType: NodeInfo.Kind) extends FNOneArg(recipes,argType) {
+  override def computeValue(arg1: DataValuePrimitive, dstate: DState): DataValuePrimitive = {
+    argType match {
+      case NodeInfo.Long => ~arg1.getLong
+      case NodeInfo.Int => ~arg1.getInt
+      case NodeInfo.Short => ~arg1.getShort
+      case NodeInfo.Byte => ~arg1.getByte
+      case NodeInfo.UnsignedLong => arg1.getBigInt.not()
```

Comment:
@stevedlawrence Is this correct for unsignedNot?

Reply:
I don't think so. The documentation for [BigInt.not() function](https://docs.oracle.com/javase/7/docs/api/java/math/BigInteger.html#not()) says:

> This method returns a negative value if and only if this BigInteger is non-negative.

The value for us will always be non-negative since it's unsigned, which means just calling `getBigInt.not()` will always created a signed value, which is definitely wrong. Instead, we need to use the ULong, with something similar to the other unsigned types. I think something like this should do it:

```scala
case NodeInfo.UnsignedLong => (~ULong(arg1.getBigInt.longValue()).toBigInt
```
So we get the BigInteger, convert it to a long, then convert that to a ULong, `not` the ULong, and then convert that back to a BigInt.

Reply:
I am not sure whether not is not working properly or is there something with the test cases.

Reply:
Yeah, I hadn't totally understood how the ULong, UInt, etc. classes work. Apparently if you bitwise-not a UShort or a UByte, it returns a UInt. I guess this is similar to how shift's also return a UInt. Which all mimics Java and it's primitive types--in Java, if you bitwise-not a Short or Byte, you get an Int. I'm not sure why Java decided that made sense, but it means we have to do things differently.

I think probably the cleanest way to do it is like this:
```scala
case UnsignedShort => UShort((~arg1.getInt).toShort).toInt
```

So we get the argument from the infoset, immediately invert it, which will get us an Int. We then cast that Int down to a Short since that's what the UShort constructor expects. We create a UShort with that and then we can get the int value, which will have all the correct bits masked off. So we're basically just using UShort to correctly mask of the bits when we convert it back to an int.

Carrying this idea to the other types, I think we want something like this:

```scala
case UnsignedLong => ULong((~arg1.getBigInt.longValue)).toBigInt
case UnsignedInt => UInt((~arg1.getLong).toInt).toLong
case UnsignedShort => UShort((~arg1.getInt).toShort).toInt
case UnsignedByte => UByte((~arg1.getShort).toByte).toShort
```

Just a side note, that if we ever update the infoset to represent these unsigned types as ULong/UInt/etc in the infoset (see [DAFFODIL-2511](https://issues.apache.org/jira/browse/DAFFODIL-2511)), then this would probably look something like this:

```scala
case UnsignedLong => ~arg1.getULong
case UnsignedInt => ~arg1.getUInt
case UnsignedShort => (~arg1.getUShort).toUShort
case UnsignedByte => (~arg1.getUByte).toUByte
```
The UnsignedShort/Byte cases looks a bit weird due to bitwise-not returning a UInt, but its significantly cleaner.

Reply:
This is also not working.
Should I remove not form the PR and merge rest of them.Not can be picked separately later.

Reply:
Probably easier to get this working than to back out changes. Can you push what you have and we can take a look?